### PR TITLE
Update codegen to support Sysl union type.

### PR DESCRIPTION
**Pull request by ericzhang6222**
This is a simple and quick fixing as #95 is a blocker issue, now it can transfer Sysl union type too. For example, transform
```
!type Cat:
        id <: string 
        age <: int?:
            @json_tag = "age"
        hunts <: bool?:
            @json_tag = "hunts"

!type Dog:
        id <: string
        bark <: bool?:
            @json_tag = "bark"
        breed <: string?:
            @json_tag = "breed"

!union PetA:
        Cat
        Dog
```
to
```
// PetA can be one of following types in runtime:
// Cat
// Dog
type PetA interface {
	// isPetA is identifier method
	isPetA()
}

// isPetA identifies Cat is instance of PetA
func (i Cat) isPetA() {
}

// isPetA identifies Dog is instance of PetA
func (i Dog) isPetA() {
}
```
It can be enhanced still, like adding util method to transform instance of `Pet` to `Cat` or `Dog`.

### CODE REVIEW

File: codegen/arrai/svc_types.arrai [hunk 1]
```diff
@@ -140,4 +141,10 @@ let fieldJsonTag = \type \key
             type ${.typename} struct {
             }
         `::\i}
+        ${unions >> \(@: key, @value: value, :typename)$`
+            // ${typename} can be one of following types in runtime:
+            ${value('oneOf')('type').a >> .('typeRef')('ref')('path').a >> ('// ' + .@item.s)::\i}
```
Comment (orlade-anz) [inline codegen/arrai/svc_types.arrai]: Nit: style
```suggestion
            ${value('oneOf')('type').a >> .('typeRef')('ref')('path').a >> $`// ${.@item.s}`::\i}
```

**Review by orlade-anz — CHANGES_REQUESTED** [review]
Generally LGTM I think, but I'm not super familiar with sysl-go yet. I'd like an approval from someone who is.

File: codegen/tests/simple/types.go [hunk 2]
```diff
@@ -18,6 +18,20 @@ var _ = time.Parse
 // Reference imports to suppress unused errors
 var _ = date.Parse
 
+// Cat ...
+type Cat struct {
```
Comment (andrewemeryanz) [inline codegen/tests/simple/types.go:22]: We should go a little further and make the relationship explicit:

```go
func (c Cat) isPet() {}
func (d Dog) isPet() {}
type Pet interface {
	isPet()
}
```

**Review by andrewemeryanz — CHANGES_REQUESTED** [review]
Overall approach good. Minor change requested to make the relationship between union types explicit.

**Review by springwiz — DISMISSED** [review]

**Review by springwiz — CHANGES_REQUESTED** [review]
I see an issue lurking over here, as per my understanding, both Cat and Dog can behave as Pet during runtime as they are part of a union(both implement Pet interface {}). But essentially in the generated code the method set associated with  Pet interface {} is empty, which means even a struct like type struct Empty {} can be a Pet. Please add a method to the Pet interface {} and implement the same in the structs Cat and Dog.

**Comment by ericzhang6222** [comment]
> I see an issue lurking over here, as per my understanding, both Cat and Dog can behave as Pet during runtime as they are part of a union(both implement Pet interface {}). But essentially in the generated code the method set associated with Pet interface {} is empty, which means even a struct like type struct Empty {} can be a Pet. Please add a method to the Pet interface {} and implement the same in the structs Cat and Dog.

Fixed.

**Review by springwiz — APPROVED** [review]
LGTM

File: codegen/tests/simple/types.go [hunk 3]
```diff
@@ -200,3 +229,39 @@ type Str string
 // Empty ...
 type Empty struct {
 }
+
+// PetA can be one of following types in runtime:
+// Cat
+// Dog
+type PetA interface {
+	// IsPetA returns true if it is instance of PetA
+	IsPetA() bool
```
Comment (andrewemeryanz) [inline codegen/tests/simple/types.go]: Method should be `isPetA()`:
1. We don't want to export the method we just want to mark the implementations.
1. We don't need to return anything; void will suffice.

**Review by andrewemeryanz — CHANGES_REQUESTED** [review]

**Review by andrewemeryanz — APPROVED** [review]

**Review by orlade-anz — APPROVED** [review]

**Pull request merged**